improve the approach we use for parsing and storing component data

superwhiskers opened the pull request:
this pull request replaces the data structures for components introduced in #133, save for those related to buses. this is done to simplify the way json parsing is handled for them

 ## checklist
 
- [x] All tests pass.
- [x] Code compiles cleanly with flags `-Wall -Wpedantic -Wconversion -Wextra`.
- [x] The new code follows GridKit™ style guidelines.
- [x] There are unit tests for the new code.
- [x] The new code is documented.
- [x] The feature branch is rebased with respect to the target branch.

superwhiskers committed:
initial progress on the new data structure parsing approach

superwhiskers committed:
Apply pre-commmit fixes

pelesh reviewed (changes requested):
Class `BusFault.cpp` includes (indirectly) both, `nlohmann/json.hpp` and `magic_enum/magic_enum.hpp` headers. This makes the component dependent on I/O file parser implementation. We need to find a way to abstract away component models from the parser specifics. All component models should build independently of the parser (where it is okay if the parser depends on components).

superwhiskers commented:
i'm not sure it's really possible to decouple it any further unless we added even more intermediary structures or something. both of those headers could be removed and with some conditional compilation, since all of the actual json parsing code resides in the one file, it'd be easy to remove the json parser and magic_enum. i don't think attempting to completely remove an indirect dependency is particularly worthwhile here.

superwhiskers commented:
well, actually now that i recall there's one field in the `ComponentData.hpp` file that needs `magic_enum`. the json parser, however, can be removed completely. it technically doesn't even have to be in that same file (but i put it in there because it simplifies things for the user as they wouldn't have to include a separate "json parser" file to use the json parser with components) as the json parser we use only requires things to be 1) in the same namespace[^1] and 2) included within the file they are being used within

[^1]: technically it's not required but it makes things simpler

pelesh commented:
Also parser tests should not be in phasor dynamics but in a separate subdirectory. They are helper utility, not the part of phasor dynamics core.

pelesh commented:
> i'm not sure it's really possible to decouple it any further unless we added even more intermediary structures or something. 

Then I guess we should add intermediary data structures, i.e. meta models. Also, we should consider using factories to instantiate and parametrize components.

superwhiskers commented:
> Then I guess we should add intermediary data structures, i.e. meta models

we already have an intermediary (the `*Data` family of structures). i'm concerned adding even more would overcomplicate the entire set of classes and data structures we have and for little tangible benefit. i already think what we're doing here is a little much, but i can see the benefit (decoupling our internal representation from the exposed one). adding even more seems unnecessary and would duplicate representations all over the place.

we can already remove the json parser from the `ComponentData` structure easily and i don't see the benefit of going any further than that as the data structure itself has no reliance upon the existence of it and it's served its purpose (making individual `*Data` structures not need to deal with json parsing and simplifying the process of adding new components).

adding even more structures would be detrimental to your original goal, i think. it would require consumers of the library to interact with more structures to get things done (unless i'm interpreting the proposal wrong) and again, duplicate places in which data can be and incur a greater maintenance burden upon us. without more justification and a clearer path to implementation i'm against making this change, personally.

superwhiskers committed:
move the case format tests and require floating point parameters to be floating point literals again. no implicit conversions

superwhiskers committed:
add some documentation to the modified bus fault data structure

superwhiskers committed:
break out the component data json parser into a separate file

superwhiskers committed:
finish reworking the remainder of the applicable `*Data` structures to use `ComponentData`

note that this doesn't build due to code relying upon the presence of fields. this will be finished in a subsequent commit

superwhiskers committed:
finish the data structure reworking started in the previous commit

superwhiskers committed:
Apply pre-commmit fixes

superwhiskers commented:
this should be ready to go now

superwhiskers committed:
fix documentation comment in `LoadImpl.hpp`

superwhiskers committed:
remove extraneous parens in `SystemModelData.hpp`

pelesh commented on CMakeLists.txt:
```diff
@@ -84,6 +84,7 @@ endif("${isSystemDir}" STREQUAL "-1")
 
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/src)
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/third-party/nlohmann-json/single_include)
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/third-party/magic-enum/include)
```
The third-party includes must not be made available to the entire code, only to the intended target(s). Please use `TARGET_INCLUDE_DIRECTORIES`.

pelesh commented on src/Model/PhasorDynamics/Load/LoadData.hpp:
```diff
@@ -6,37 +6,55 @@
+      R,
+
+      /// Load reactance
+      X,
```
Please use compact list of parameters for all components:
```suggestion
      R, ///< Load resistance
      X, ///< Load reactance
```
It is easier to see what parameters are available for each model. In most cases component model developers would recognize the parameter by its name, so a brief comment on the side is sufficient.

superwhiskers committed:
fix the `ThreeBusClassical` example

superwhiskers committed:
Apply pre-commmit fixes

superwhiskers commented:
rebased on latest develop and made changes to fix the new `ThreeBusClassical` test

superwhiskers committed:
minor fixes to the documentation comments, remove unnecessary no-ops

superwhiskers committed:
Apply pre-commmit fixes

pelesh reviewed (commented):
Just a couple of nitpicking comments.

superwhiskers committed:
address all review comments

superwhiskers committed:
Apply pre-commmit fixes
